src/NextGen/Framework/Receipts/Actions/GenerateConfirmationPageReceipt.php

```diff
@@ -0,0 +1,176 @@
+<?php
+
+namespace Give\NextGen\Framework\Receipts\Actions;
+
+use Give\Donations\Models\Donation;
+use Give\Framework\FieldsAPI\Concerns\HasLabel;
+use Give\Framework\FieldsAPI\Concerns\HasName;
+use Give\Framework\FieldsAPI\Field;
+use Give\NextGen\DonationForm\Models\DonationForm;
+use Give\NextGen\DonationForm\Repositories\DonationFormRepository;
+use Give\NextGen\Framework\Receipts\DonationReceipt;
+use Give\NextGen\Framework\Receipts\Properties\ReceiptDetail;
+
+class GenerateConfirmationPageReceipt
+{
+    /**
+     * @unreleased
+     */
+    public function __invoke(DonationReceipt $receipt): DonationReceipt
+    {
+        $this->fillSettings($receipt);
+        $this->fillDonorDetails($receipt);
+        $this->fillDonationDetails($receipt);
+        $this->fillSubscriptionDetails($receipt);
+        $this->fillAdditionalDetails($receipt);
+
+        return $receipt;
+    }
+
+    /**
+     * @unreleased
+     */
+    protected function getCustomFields(Donation $donation): array
+    {
+        if (give(DonationFormRepository::class)->isLegacyForm($donation->formId)) {
+            return [];
+        }
+
+        /** @var DonationForm $form */
+        $form = DonationForm::find($donation->formId);
+
+        $customFields = array_filter($form->schema()->getFields(), static function (Field $field) {
+            /** $field->shouldDisplayInReceipt is a temporary macro */
+            return $field->shouldDisplayInReceipt();
+        });
+
+        return array_map(static function (Field $field) use ($donation) {
+            /** @var Field|HasLabel|HasName $field */
+            return new ReceiptDetail(
+                $field->getLabel(),
+                give()->payment_meta->get_meta($donation->id, $field->getName(), true)
+            );
+        }, $customFields);
+    }
+
+    /**
+     * @unreleased
+     *
+     * @return void
+     */
+    private function fillDonationDetails(DonationReceipt $receipt)
+    {
+        $receipt->donationDetails->addDetails([
+                new ReceiptDetail(
+                    __('Payment Status', 'give'),
+                    give_get_payment_statuses()[$receipt->donation->status->getValue()]
```

Comment:
Instead of reaching into the `DonationReciept` to access the `Donation` I think we should pass the `Donation` separately - generating a receipt that is filled with the donation data.

Reply:
Reaching into the `DonationReciept` to access the `Donation` tightly couples the shape of the `Donation` object as a de facto API of the receipt, which is mixing concerns. Therefor, changes to the `Donation` result in unintentional changes to the `DonationReciept`.

Reply:
@kjohnson I hear what you're saying but i'm also on the fence about this.  Ultimately, a donation receipt is nothing without a donation, it represents the donor facing data related to a donation.  In that regard, the `DonationReceipt` object in my mind should contain a donation model instance to even be created.  I like the simplicity of that relationship as it works downstream for generating a receipt and modifying one since you can access the underlying donation model from the receipt itself.

Reply:
@jonwaldstein that is fair. I'm happy to lay this discussion to rest. 👍🏻